File: G. Monads/safeRoot.py
from math import sqrt, pow
from typing import Callable
from functools import reduce
import operator
from pymonad.Maybe import *

# Dr. Abbott's cleaned up Infix "hack" from the forum
# Uses the two pipes to create an infix work around
class Infix:
    def __init__(self, function):
        self.fn = function

    def __ror__(self, left_arg):
        """
        For | on left:
        Store the left argument as self.left_arg
        and return this object.
        """
        self.left_arg = left_arg
        return self

    def __or__(self, right_arg):
        """For | on right: Execute the function. """
        return self.fn(self.left_arg, right_arg)

# Square root function that uses the Maybe monad
def safeSqrt(x: int) -> Maybe:
    if x < 0:
        return Nothing
    else:
        return Just(sqrt(x))

# Uses PyMonad's bind function (>>) to call safeSqrt twice on a number
def testSafeSqrt0(x: int) -> Maybe:
    return safeSqrt(x) >> safeSqrt

# Another more convoluted/explicit way of doing testSafeSqrt0
def testSafeSqrt1(x: int) -> Maybe:
    if safeSqrt(x) is not Nothing:
        return safeSqrt(safeSqrt(x).value)
    else:
        return Nothing

# Test for equivalence between two tests above:

# Checks if all in comprehension checking for equivalence are true
# testEq0 a b = all (\x -> testSafeSqrt0 x == testSafeSqrt1 x) [a .. b]
def testEq0(a: int, b: int) -> bool:
    return all(testSafeSqrt0(x) == testSafeSqrt1(x) for x in range(a,b + 1))

# Checks if the values from the Maybe monad are equivalent using a sort of ID function
# testEq1 a b = all id [testSafeSqrt0 x == testSafeSqrt1 x | x <- [a .. b]]
def testEq1(a: int, b: int) -> bool:
    return all(x == x for x in [testSafeSqrt0(x).value == testSafeSqrt1(x).value for x in range(a,b + 1)])

# Uses reduce function (function, sequence, initial) on the and operator and a list comprehension of bools
# testEq2 a b = foldl (&&) True [testSafeSqrt0 x == testSafeSqrt1 x | x <- [a .. b]]
def testEq2(a: int, b: int) -> bool:
    return reduce(operator.and_, [testSafeSqrt0(x) == testSafeSqrt1(x) for x in range(a,b + 1)], True)

print("testEq0: ", testEq0(-5,5))
print("testEq1: ", testEq1(-5,5))
print("testEq2: ", testEq2(-5,5))


# Composition function used by Infix hack.
def comp(f1: Callable, f2: Callable):
    return lambda x: f2(x) >> f1

# Declare variable to be used as "operator" in Infix
c = Infix(comp)

# safeRoot works recursively to return a function. The base case (when n is 0)
# returns the function Just. Otherwise it calls itself on the decremented value.
# When the base case is reached Just is applied to 2^2^0 = 2
# It basically builds up a stack of n-1 safeSqrt calls on the initial 2^2^n value,
# always returning Just 2.
def safeRoot(n: int) -> Callable[[float], Maybe]:
    if n is 0:
        return Just
    else:
        return safeSqrt |c| safeRoot(n-1)

# helper function that calls safeRoot
def testSafeRoot(n: int) -> Maybe:
    return safeRoot(n)(pow(2, pow(2, n)))

# tester function that checks if all values return from calling testSafeRoot on
# 0-9 are Just(2)
def testSafeRootTo9():
    print("testSafeRootTo9: ", all(x == Just(2) for x in [testSafeRoot(n) for n in range(0,10)]))

testSafeRootTo9()

File: G. Monads/safeRoot.hs

-- safeSqrt returns Just (sqrt x) if x >= 0. If x < 0, it returns Nothing.
safeSqrt :: Double -> Maybe Double
safeSqrt x
    | x < 0     = Nothing
    | otherwise = Just (sqrt x)


-- Two equivalent ways to execute 2 safeSqrt calls in sequence.
testSafeSqrt0 :: Double -> Maybe Double
testSafeSqrt0 x = safeSqrt x >>= safeSqrt

testSafeSqrt1 :: Double -> Maybe Double
testSafeSqrt1 x = do
                     y <- safeSqrt x
                     safeSqrt y

-- Three tests (all equivalent) to ensure testSafeSqrt0 and testSafeSqrt1 are the same
testEq0 a b = all (\x -> testSafeSqrt0 x == testSafeSqrt1 x) [a .. b]

testEq1 a b = all id [testSafeSqrt0 x == testSafeSqrt1 x | x <- [a .. b]]

testEq2 a b = foldl (&&) True [testSafeSqrt0 x == testSafeSqrt1 x | x <- [a .. b]]

-- > testEq0 (-5) 5
-- True


-- Define a function that composes two "safe" functions.
-- This is slightly different from >>=.
--
-- The type of >>= is
-- (>>=) :: Monad m => m b -> (b -> m c) -> m c
-- which takes a wrapped value and a function.
-- Here we will define a function that takes two functions and produces
-- their composition.
--
-- First let's define an infix operator (<.>) for the function. Defining it with
-- infixr 1 <.>   declares it to associate to the right and to have precedence 1.
-- Notice that (<.>) takes two functions as arguments. It returns a function
-- that takes an argument to the first function and returns the output
-- of the two functions in sequence, with the appropriate unwrapping between them.
infixr 1 <.>
(<.>) :: Monad m => (a -> m b) -> (b -> m c) -> (a -> m c)
-- Two definitions: in do notation and using explicit >>=.
-- Either one will do the job.
-- f1 <.> f2 = \x -> do y <- f1 x
--                      f2 y
f1 <.> f2 = \x -> f1 x >>= f2

-- Let's create a function that builds a safe function that takes the
-- (1/2)^n the root of its argument.
-- safeRoot 0 == Take no roots; wrap the argument in Just.
--                         (safeRoot 0) (2^1) => Just 2.0
-- safeRoot 1 == safeSqrt; (safeRoot 1) (2^2) => Just 2.0
-- safeRoot 2 == safeFourthRoot; (safeRoot 2) (2^4) => Just 2.0
-- safeRoot 3 == safeEighthRoot; (safeRoot 3) (2^8) => Just 2.0
-- safeRoot 4 == safeSixteenthRoot; (safeRoot 4) (2^16) => Just 2.0
-- safeRoot n == safeNthRoot; (safeRoot n) (2^(2^n)) => Just 2.0
-- Biuld the new function by recursive composition with safeSqrt.
-- The parentheses are not necessary since -> associates to the right.
-- They are inteded to make it clear that safeRoot takes an Int and
-- returns a function from Double to Maybe Double.
safeRoot :: Int -> (Double -> Maybe Double)
safeRoot n
    | n == 0 = Just  -- Just is a function, a type constructor. Just :: a -> Maybe a
    | otherwise = safeSqrt <.> safeRoot (n - 1)

-- Test safeRoot by asking it to generate a function (safeRoot n), which
-- when given an input should return Just 2.0
testSafeRoot :: Int -> Maybe Double
testSafeRoot n = (safeRoot n) (2^(2^n)) -- Should return Just 2.0 for any n.

-- > testSafeRoot 5
-- Just 2.0

-- Test for multiple n's. Limit to a maximum value of 9 becaues 2^(2^10)
-- is too large for a Double. Haskell can represent 2^(2^10) as an Integer,
-- but (safeRoot n) is a function that expects a Double as input.
testSafeRootTo9 :: Bool
testSafeRootTo9 = all (== Just 2) [testSafeRoot n | n <- [0 .. 9]]

-- > testSafeRootTo9
-- True
